src/cocaine-app/history.py

```diff
@@ -0,0 +1,401 @@
+        group_history_list = GroupHistory.list(
+            self.collection,
+            group_id=group_id).limit(1)
+        if not group_history_list.count():
+            raise GroupHistoryNotFoundError('Group history for group {0} is not found'.format(
+                group_id
+            ))
+        group_history = GroupHistory.new(**group_history_list[0])
+        group_history.collection = self.collection
+        return group_history
+
+    def group_history(self, group_id):
+        try:
+            return self._get_group_history(group_id)
+        except GroupHistoryNotFoundError:
+            return None
+        except Exception:
+            logger.exception('History finder failed')
+            return None
+
+    def groups_history(self, ids=None):
+        jobs = [
+            GroupHistory(gh) for gh in GroupHistory.list(
+                self.collection,
+                group_id=ids
+            )
+        ]
+        for j in jobs:
+            j.collection = self.collection
+            j._dirty = False
+        return jobs
+
+    def search_by_node_backend(self,
+                               hostname=None,
+                               port=None,
+                               family=None,
+                               backend_id=None,
+                               path=None):
+        node_backend_pattern = {}
+        if hostname:
+            node_backend_pattern['hostname'] = hostname
+        if port:
+            node_backend_pattern['port'] = port
+        if family:
+            node_backend_pattern['family'] = family
+        if backend_id:
+            node_backend_pattern['backend_id'] = backend_id
+        if path:
+            if not path.startswith('^'):
+                path = '^' + path
+            if not path.endswith('$'):
+                path = path + '$'
+            path = path.replace('*', '.*')
+            node_backend_pattern['path'] = {'$regex': path}
+
+        if not node_backend_pattern:
+            raise ValueError('Filter parameters for node backends should be supplied')
+
+        group_histories_records = self.collection.find({
+            'nodes.set': {
+                '$elemMatch': node_backend_pattern
+            }
+        })
+
+        ghs = []
+        for group_history_record in group_histories_records:
+            gh = GroupHistory(**group_history_record)
+            gh.collection = self.collection
+            ghs.append(gh)
+        return ghs
+
+    def search_by_history_record(self,
+                                 start_ts=None,
+                                 finish_ts=None,
+                                 type=None):
+        record_pattern = {}
+        if type:
+            record_pattern.update(condition('type', type))
+        if start_ts or finish_ts:
+            timestamp_pattern = {}
+            if start_ts:
+                timestamp_pattern['$gte'] = start_ts
+            if finish_ts:
+                timestamp_pattern['$lt'] = finish_ts
```

Comment:
Because that point will be overlapped through consecutive calls. The idea is to work like this:
```
[ts1, ts2)
[ts2, ts3)
...
```